apps/system/utils/kdbg_cpuload.c

```diff
@@ -0,0 +1,207 @@
+#include <string.h>
+#include <sched.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <tinyara/clock.h>
+#include <tinyara/sched.h>
+#include "kdbg_utils.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define CPULOADMON_PREFIX "CPUload Monitor: "
+
+#define CPULOADMONITOR_STACKSIZE 1024
+
+#ifndef CONFIG_CPULOADMONITOR_PRIORITY
+#define CONFIG_CPULOADMONITOR_PRIORITY 100
+#endif
+
+#ifndef CONFIG_CPULOADMONITOR_INTERVAL
+#define CONFIG_CPULOADMONITOR_INTERVAL 5
+#endif
+
+extern volatile uint32_t g_cpuload_timeconstant[SCHED_NCPULOAD];
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static volatile bool is_started = false;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void cpuload_print_task(struct tcb_s *tcb, void *arg)
+{
+	struct cpuload_s cpuload;
+	uint32_t intpart;
+	uint32_t fracpart;
+	int cpuload_idx;
+	uint32_t tmp;
+
+	printf("  %d    ", tcb->pid);
+	for (cpuload_idx = 0; cpuload_idx < SCHED_NCPULOAD; cpuload_idx++) {
+
+		(void)clock_cpuload(tcb->pid, cpuload_idx, &cpuload);
+
+		if (cpuload.total > 0) {
+			tmp = (1000 * cpuload.active) / cpuload.total;
+			intpart = tmp / 10;
+			fracpart = tmp - 10 * intpart;
+		} else {
+			intpart = 0;
+			fracpart = 0;
+		}
+		printf("%3d.%d %%   ", intpart, fracpart);
+	}
+	printf("\n");
+}
+
+static void cpuload_print_tasklist(void)
+{
+	int cpuload_idx;
+
+#ifdef CONFIG_SCHED_MULTI_CPULOAD
+	printf("=====================================\n");
+#else
+	printf("=================\n");
+#endif
+	printf(" CPU USAGE\n");
+#ifdef CONFIG_SCHED_MULTI_CPULOAD
+	printf("=====================================\n");
+#else
+	printf("=================\n");
+#endif
+	printf(" PID");
+	for (cpuload_idx = 0; cpuload_idx < SCHED_NCPULOAD; cpuload_idx++) {
+		printf("%9ds", g_cpuload_timeconstant[cpuload_idx]);
+	}
+#ifdef CONFIG_SCHED_MULTI_CPULOAD
+	printf("\n-------------------------------------\n");
+#else
+	printf("\n-----------------\n");
+#endif
+
+	/* Print cpu load for each task */
+	sched_foreach(cpuload_print_task, NULL);
+
+#ifdef CONFIG_SCHED_MULTI_CPULOAD
+	printf("=====================================\n");
+#else
+	printf("=================\n");
+#endif
+}
+
+static void *cpuloadmonitor_daemon(void *args)
+{
+	/* Loop until we detect that there is a request to stop. */
+	while (is_started) {
+		cpuload_print_tasklist();		
+		sleep(CONFIG_CPULOADMONITOR_INTERVAL);
+	}
+
+	/* Stopped */
+	printf(CPULOADMON_PREFIX "Stopped well\n");
+
+	return NULL;
+}
+
+static void cpuload_daemon_start(void)
+{
+	int ret;
+	pthread_t cpuloadmon;
+	pthread_attr_t attr;
+
+	if (is_started) {
```

Comment:
Even though is_started is false by calling cpuload_daemon_stop(), cpuloadmon might be still alive.
Unfortunately, more than two cpu load monitors are created.
Need more logic to handle this.